File: fsharp/error-handling/ErrorHandling.fs
module ErrorHandling

let handleErrorByThrowingException () = failwith "Here's your exception"

let handleErrorByReturningOption (input: string) =
    let mutable value = 0

    if System.Int32.TryParse(input, &value) then
        Some value
    else
        None

let handleErrorByReturningResult (input: string) =
    let mutable value = 0

    if System.Int32.TryParse(input, &value) then
        Ok value
    else
        Error "Could not convert input to integer"

let bind = Result.bind

let cleanupDisposablesWhenThrowingException resource =
    use r = resource
    failwith "Oh no!"

File: csharp/error-handling/ErrorHandling.cs
using System;

public static class ErrorHandling
{
    public static void HandleErrorByThrowingException() => 
        throw new Exception("Hey, you asked for it.");

    public static int? HandleErrorByReturningNullableType(string input)
    {
        if (int.TryParse(input, out var value))
        {
            return value;
        }

        return null;
    }

    public static bool HandleErrorWithOutParam(string input, out int result) => 
        int.TryParse(input, out result);

    public static void DisposableResourcesAreDisposedWhenExceptionIsThrown(IDisposable disposableObject)
    {
        using var resource = disposableObject;
        throw new Exception("Hi, I'm an error.");
    }
}
